File: ddns_netcup/netcup.py
from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Dict, List, Optional, Tuple
from urllib import request

API_URL = "https://ccp.netcup.net/run/webservice/servers/endpoint.php?JSON"


class NetcupAPIError(RuntimeError):
    pass


@dataclass
class NetcupCredentials:
    customer_number: str
    api_key: str
    api_password: str


class NetcupClient:
    def __init__(self, credentials: NetcupCredentials, timeout: int = 30):
        self.credentials = credentials
        self.timeout = timeout

    def _request(self, payload: Dict) -> Dict:
        body = json.dumps(payload).encode("utf-8")
        req = request.Request(
            API_URL,
            data=body,
            method="POST",
            headers={"Content-Type": "application/json", "User-Agent": "ddns-netcup-bridge/1.0"},
        )
        with request.urlopen(req, timeout=self.timeout) as resp:
            raw = resp.read()

        parsed = json.loads(raw)
        if parsed.get("status") != "success":
            raise NetcupAPIError(parsed.get("longmessage", "Unknown Netcup API error"))
        return parsed

    def _auth_payload(self, apisessionid: Optional[str] = None) -> Dict:
        payload = {
            "customernumber": self.credentials.customer_number,
            "apikey": self.credentials.api_key,
        }
        if apisessionid is None:
            payload["apipassword"] = self.credentials.api_password
        else:
            payload["apisessionid"] = apisessionid
        return payload

    def login(self) -> str:
        result = self._request({"action": "login", "param": self._auth_payload()})
        return result["responsedata"]["apisessionid"]

    def logout(self, session_id: str) -> None:
        self._request({"action": "logout", "param": self._auth_payload(session_id)})

    def info_dns_records(self, domain: str, session_id: str) -> List[Dict]:
        result = self._request(
            {
                "action": "infoDnsRecords",
                "param": {
                    **self._auth_payload(session_id),
                    "domainname": domain,
                },
            }
        )
        records = result["responsedata"]["dnsrecords"]
        if not isinstance(records, list):
            raise NetcupAPIError("Invalid DNS record response")
        return records

    def update_dns_records(self, domain: str, records: List[Dict], session_id: str) -> None:
        self._request(
            {
                "action": "updateDnsRecords",
                "param": {
                    **self._auth_payload(session_id),
                    "domainname": domain,
                    "dnsrecordset": {"dnsrecords": records},
                },
            }
        )


class NetcupDNSService:
    def __init__(self, client: NetcupClient):
        self.client = client

    @staticmethod
    def _apply_record(records: List[Dict], host: str, record_type: str, value: str) -> bool:
        changed = False
        matched = False
        for record in records:
            if record.get("hostname") == host and record.get("type") == record_type:
                matched = True
                if record.get("destination") != value:
                    record["destination"] = value
                    changed = True

        if not matched:
            records.append(
                {
                    "hostname": host,
                    "type": record_type,
                    "priority": "0",
                    "destination": value,
                    "deleterecord": "false",
                    "state": "yes",
                }
            )
            changed = True

        return changed

    def update_targets(self, targets: List[Tuple[str, str]], ipv4: Optional[str], ipv6: Optional[str]) -> bool:
        any_changed = False
        session_id = self.client.login()
        try:
            by_domain: Dict[str, List[str]] = {}
            for domain, host in targets:
                by_domain.setdefault(domain, []).append(host)

            for domain, hosts in by_domain.items():
                records = self.client.info_dns_records(domain, session_id)
                changed = False
                for host in hosts:
                    if ipv4:
                        changed = self._apply_record(records, host, "A", ipv4) or changed
                    if ipv6:
                        changed = self._apply_record(records, host, "AAAA", ipv6) or changed
                if changed:
                    self.client.update_dns_records(domain, records, session_id)
                    any_changed = True
        finally:
            self.client.logout(session_id)

        return any_changed

File: src/netcup.ts
/**
 * Low-level client for the Netcup CCP DNS API (JSON variant).
 *
 * Endpoint: https://ccp.netcup.net/run/webservice/servers/endpoint.php?JSON
 * Every request is an HTTP POST with body `{ action, param }`. The session is
 * established with `login` (returns an `apisessionid`) and torn down with
 * `logout`.
 *
 * This module intentionally contains no business logic — it just speaks the
 * protocol. The diff/decision logic lives in updater.ts so it can be unit
 * tested without the network.
 */

import type { NetcupConfig } from "./config.js";
import { logger } from "./logger.js";

const REQUEST_TIMEOUT_MS = 15_000;
/** Netcup status code returned when a zone simply has no records yet. */
const STATUS_NO_RECORDS = 5029;

export interface DnsRecord {
  id?: string;
  hostname: string;
  type: string;
  priority?: string;
  destination: string;
  deleterecord?: boolean;
  state?: string;
}

interface ApiResponse<T> {
  serverrequestid: string;
  action: string;
  status: "success" | "error" | "started" | "pending";
  statuscode: number;
  shortmessage: string;
  longmessage: string;
  responsedata: T;
}

export class NetcupError extends Error {
  constructor(
    readonly action: string,
    readonly statuscode: number,
    message: string,
  ) {
    super(`Netcup ${action} failed (${statuscode}): ${message}`);
    this.name = "NetcupError";
  }
}

export class NetcupClient {
  constructor(private readonly config: NetcupConfig) {}

  private async call<T>(action: string, param: Record<string, unknown>): Promise<ApiResponse<T>> {
    let res: Response;
    try {
      res = await fetch(this.config.apiUrl, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ action, param }),
        signal: AbortSignal.timeout(REQUEST_TIMEOUT_MS),
      });
    } catch (err) {
      throw new NetcupError(action, 0, `network error: ${(err as Error).message}`);
    }

    const data = (await res.json()) as ApiResponse<T>;
    if (data.status !== "success") {
      throw new NetcupError(action, data.statuscode, data.longmessage || data.shortmessage);
    }
    return data;
  }

  private async login(): Promise<string> {
    const res = await this.call<{ apisessionid: string }>("login", {
      customernumber: this.config.customerNumber,
      apikey: this.config.apiKey,
      apipassword: this.config.apiPassword,
    });
    return res.responsedata.apisessionid;
  }

  private async logout(apisessionid: string): Promise<void> {
    try {
      await this.call("logout", {
        customernumber: this.config.customerNumber,
        apikey: this.config.apiKey,
        apisessionid,
      });
    } catch (err) {
      // A failed logout is non-fatal; just note it.
      logger.warn("Netcup logout failed", { error: (err as Error).message });
    }
  }

  private async infoDnsRecords(apisessionid: string, domain: string): Promise<DnsRecord[]> {
    try {
      const res = await this.call<{ dnsrecords: DnsRecord[] }>("infoDnsRecords", {
        domainname: domain,
        customernumber: this.config.customerNumber,
        apikey: this.config.apiKey,
        apisessionid,
      });
      return res.responsedata.dnsrecords ?? [];
    } catch (err) {
      // An empty zone is reported as an error code; treat it as "no records".
      if (err instanceof NetcupError && err.statuscode === STATUS_NO_RECORDS) return [];
      throw err;
    }
  }

  private async updateDnsRecords(
    apisessionid: string,
    domain: string,
    records: DnsRecord[],
  ): Promise<DnsRecord[]> {
    const res = await this.call<{ dnsrecords: DnsRecord[] }>("updateDnsRecords", {
      domainname: domain,
      customernumber: this.config.customerNumber,
      apikey: this.config.apiKey,
      apisessionid,
      // Only the records we want to change are sent — no full-zone rewrite.
      dnsrecordset: { dnsrecords: records },
    });
    return res.responsedata.dnsrecords ?? [];
  }

  /**
   * Run a unit of work against a single domain inside one authenticated
   * session. `work` receives the current records and returns the (possibly
   * empty) set of records to push back. Logout always runs.
   */
  async withDomain(
    domain: string,
    work: (records: DnsRecord[]) => DnsRecord[],
  ): Promise<void> {
    const session = await this.login();
    try {
      const existing = await this.infoDnsRecords(session, domain);
      const toUpdate = work(existing);
      if (toUpdate.length > 0) {
        await this.updateDnsRecords(session, domain, toUpdate);
      }
    } finally {
      await this.logout(session);
    }
  }
}
